contracts/RelayHub.sol

```diff
@@ -187,42 +182,54 @@ contract RelayHub is IRelayHub {
         require(externalGasLimit <= block.gaslimit, "Impossible gas limit");
 
         // We now verify that the paymaster will agree to be charged for the transaction.
-        (vars.success, vars.recipientContext, vars.gasLimits) =
-            canRelay(relayRequest,
-                    externalGasLimit, signature, approvalData);
-
-        if (!vars.success) {
-            revertReason = GsnUtils.getError(vars.recipientContext);
-            emit TransactionRejectedByPaymaster(
-                workerToManager[msg.sender],
-                relayRequest.relayData.paymaster,
-                relayRequest.request.from,
-                relayRequest.request.to,
-                msg.sender,
-                vars.functionSelector,
-                revertReason);
-            return (vars.success, revertReason);
-        }
-
-        // From this point on, this transaction will not revert nor run out of gas, and the paymaster will be charged
-        // for the gas spent.
+        (vars.gasLimits, vars.maxPossibleGas) =
+            checkGasLimits(relayRequest, externalGasLimit);
 
     {
+        vars.externalGasLimit = externalGasLimit;
+
         //How much gas to pass down to innerRelayCall. must be lower than the default 63/64
         // actually, min(gasleft*63/64, gasleft-GAS_RESERVE) might be enough.
         uint innerGasLimit = gasleft()*63/64-gasReserve;
+        vars.gasBeforeInner = gasleft();
 
         // Calls to the recipient are performed atomically inside an inner transaction which may revert in case of
         // errors in the recipient. In either case (revert or regular execution) the return data encodes the
         // RelayCallStatus value.
-        (, bytes memory relayCallStatus) = address(this).call{gas:innerGasLimit}(
-            abi.encodeWithSelector(RelayHub.innerRelayCall.selector, relayRequest, signature, vars.gasLimits,
-                innerGasLimit + externalGasLimit-gasleft() + gasOverhead + postOverhead, /*totalInitialGas*/
-                abi.decode(vars.recipientContext, (bytes)))
+        (bool success, bytes memory relayCallStatus) = address(this).call{gas:innerGasLimit}(
+            abi.encodeWithSelector(RelayHub.innerRelayCall.selector, relayRequest, signature, approvalData, vars.gasLimits,
+                innerGasLimit + vars.externalGasLimit - gasleft() + gasOverhead + postOverhead, /*totalInitialGas*/
```

Comment:
in the outer call, we do: externalGasLimit-gasLeft() + gasOverhead
here we want to do the same, with 2 exceptions:
- we don't know how much postRelayedCall will cost, so we dont' want to have it in the calculation
- we're one level deeper in the stack, so we want to add also the gas usage just before the call
  - externalGasLimit-gasLeft() - this is the gas used so far.
  - innerGasLimit-gasleft() (when gasleft() called INSIDE InternalRelayedCall) - is gas used by internalRelayedCall

so we just sum them up, and send a single var (totalInitialGas) into innerRelayedCall)

the overhead can be replaced with a single prePostOveread, instead of summing them up each time (maybe it will be clearer, since the relation of the 2 overheads is not clear here)